### PR TITLE
Ins 22 logic runner

### PR DESCRIPTION
**Pull request by Vany**
Partially implement LogicRunner interface called GoPlugin.
It can run go code formed as plugins

code have tests

### CODE REVIEW

File: logicrunner/goplugin/ginsider/ginsider.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+// GoInsider is an RPC interface to run code of plugins
+type GoInsider struct {
+	dir        string
+	RPCAddress string
+}
+
+// NewGoInsider creates a new GoInsider instance validating arguments
+func NewGoInsider(path string) *GoInsider {
```
Comment (golangci) [inline logicrunner/goplugin/ginsider/ginsider.go]: `NewGoInsider` is unused

File: logicrunner/goplugin/testplugins/secondary/main.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+func (hw *HelloWorlder) Echo(s string) (string, error) {
+	hw.Greeted++
+	return s, nil
+}
+
+func (hw HelloWorlder) ConstEcho(s string) (string, error) {
+	return s, nil
+}
+
+func JustExportedStaticFunction(int, int) {}
```
Comment (golangci) [inline logicrunner/goplugin/testplugins/secondary/main.go]: `JustExportedStaticFunction` is unused

File: logicrunner/goplugin/testplugins/secondary/main.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+var INSEXPORT HelloWorlder
```
Comment (golangci) [inline logicrunner/goplugin/testplugins/secondary/main.go]: `INSEXPORT` is unused

File: logicrunner/goplugin/ginsider/ginsider.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	rpc.HandleHTTP()
+	listener, err := net.Listen("tcp", *listen)
+
+	log.Print("ginsider launched, listens " + *listen)
+	if err != nil {
+		log.Fatal("listen error:", err)
+		os.Exit(1)
+	}
+	go http.Serve(listener, nil)
```
Comment (golangci) [inline logicrunner/goplugin/ginsider/ginsider.go]: Error return value of `http.Serve` is not checked

File: logicrunner/goplugin/goplugin.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	gp.runner = runner
+	go gp.Start()
+	return &gp, nil
+}
+
+// Start starts runner and RPC interface to help runner, note that NewGoPlugin does
+// this for you
+func (gp *GoPlugin) Start() {
+	r := GoPluginRPC{gp: gp}
+	rpc.Register(&r)
```
Comment (golangci) [inline logicrunner/goplugin/goplugin.go]: Error return value of `rpc.Register` is not checked

File: logicrunner/goplugin/goplugin.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	rpc.HandleHTTP()
+	l, e := net.Listen("tcp", gp.Options.Listen)
+	if e != nil {
+		log.Fatal("listen error:", e)
+	}
+	gp.sock = l
+	log.Printf("START")
+	http.Serve(l, nil)
```
Comment (golangci) [inline logicrunner/goplugin/goplugin.go]: Error return value of `http.Serve` is not checked

File: logicrunner/goplugin/goplugin.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	log.Printf("STOP")
+}
+
+// Stop stops runner(s) and RPC service
+func (gp *GoPlugin) Stop() {
+	gp.runner.Process.Kill()
```
Comment (golangci) [inline logicrunner/goplugin/goplugin.go]: Error return value of `gp.runner.Process.Kill` is not checked

File: logicrunner/goplugin/goplugin_test.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+)
+
+type HelloWorlder struct {
+	Greeted int
+}
+
+func compileBinaries() error {
+	d, _ := os.Getwd()
+
+	os.Chdir(d + "/ginsider")
```
Comment (golangci) [inline logicrunner/goplugin/goplugin_test.go]: Error return value of `os.Chdir` is not checked

File: logicrunner/goplugin/goplugin_test.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+	err := exec.Command("go", "build", "ginsider.go").Run()
+	if err != nil {
+		return errors.Wrap(err, "can't build ginsider")
+	}
+
+	os.Chdir(d + "/testplugins")
```
Comment (golangci) [inline logicrunner/goplugin/goplugin_test.go]: Error return value of `os.Chdir` is not checked

File: logicrunner/goplugin/goplugin_test.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	err = exec.Command("make", "secondary.so").Run()
+	if err != nil {
+		return errors.Wrap(err, "can't build pluigins")
+	}
+
+	os.Chdir(d)
```
Comment (golangci) [inline logicrunner/goplugin/goplugin_test.go]: Error return value of `os.Chdir` is not checked

File: logicrunner/goplugin/goplugin_test.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+		},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer gp.Stop()
+
+	var buff bytes.Buffer
+	e := cbor.NewEncoder(&buff)
+	e.Marshal(HelloWorlder{77})
```
Comment (golangci) [inline logicrunner/goplugin/goplugin_test.go]: Error return value of `e.Marshal` is not checked

File: logicrunner/goplugin/preprocessor/main.go [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+func init() {
+	flag.Parse()
+}
+
+func main() {
+	log.Println(os.Getwd())
+	for _, fn := range flag.Args() {
+		w := generateForFile(fn)
+		io.Copy(os.Stdout, w)
```
Comment (golangci) [inline logicrunner/goplugin/preprocessor/main.go]: Error return value of `io.Copy` is not checked

File: logicrunner/goplugin/testplugins/secondary/main.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
+package main
+
+import "errors"
+
+// @inscontract
+type HelloWorlder struct { //Docccc
```
Comment (golangci) [inline logicrunner/goplugin/testplugins/secondary/main.go]: type HelloWorlder is unused

**Comment by Vany** [comment]
Looks Good

**Pull request merged**